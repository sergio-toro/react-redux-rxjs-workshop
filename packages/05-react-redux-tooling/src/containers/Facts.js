// @flow
import { connect } from "react-redux";

import { fetchFacts, fetchFactsSuccess, fetchFactsError } from "../modules/facts";

import Facts from "../components/Facts";

const mapStateToProps = state => ({
  hasError: state.facts.hasError,
  isLoading: state.facts.isLoading,
  facts: state.facts.list
});

const mapDispatchToProps = dispatch => ({
  async onFetchFacts() {
    try {
      dispatch(fetchFacts());
      const response = await fetch(`https://kodify-workshop.netlify.com/.netlify/functions/facts?amount=6`);
      if (response.status === 200) {
        const facts = await response.json();
        dispatch(fetchFactsSuccess(facts));
      }
    } catch (error) {
      dispatch(fetchFactsError());
    }
  }
});

const FactsContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(Facts);

export default FactsContainer;
