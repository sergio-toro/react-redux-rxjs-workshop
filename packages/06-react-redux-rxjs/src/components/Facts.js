// @flow
import React, { Component, Fragment } from "react";
import styled from "styled-components";

import { ErrorMessage, Button as BaseButton, Fact as BaseFact, Loading as BaseLoading } from "00-components";

const Button = styled(BaseButton)`
  margin-left: 0;
  margin-right: 20px;
`;

const Results = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  padding-top: 20px;
`;

const Fact = styled(BaseFact)`
  margin-bottom: 10px;
  flex-basis: 100%;

  @media (min-width: 1024px) {
    flex-basis: calc(50% - 5px);
  }
`;

const Loading = styled(BaseLoading)`
  margin: 50px auto 0;
`;

type Props = {
  hasError: boolean,
  isLoading: boolean,
  facts: Object[]
};

class Facts extends Component<Props> {
  componentDidMount() {
    this.props.onFetchFacts();
  }

  render() {
    const { hasError, isLoading, facts, onFetchFacts } = this.props;
    return (
      <Fragment>
        <header>
          <h1>Cat facts</h1>
        </header>

        <Button onClick={onFetchFacts}>Refresh Facts</Button>

        <Results>
          {hasError && <ErrorMessage>Error while fetching facts, try again later</ErrorMessage>}
          {isLoading && <Loading size={100} />}
          {!isLoading && facts.map(fact => <Fact key={fact._id} text={fact.text} type={fact.type} />)}
        </Results>
      </Fragment>
    );
  }
}

export default Facts;
