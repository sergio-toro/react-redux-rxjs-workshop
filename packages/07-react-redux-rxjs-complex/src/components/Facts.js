// @flow
import React, { Component, Fragment } from "react";
import styled from "styled-components";

import {
  ErrorMessage,
  Button as BaseButton,
  Fact as BaseFact,
  Loading as BaseLoading,
  SelectList as BaseSelectList,
  RadioList as BaseRadioList,
  CheckboxList as BaseCheckboxList
} from "00-components";

const Button = styled(BaseButton)`
  margin-left: 0;
  margin-right: 20px;
`;

const RadioList = styled(BaseRadioList)`
  padding-right: 20px;
  line-height: 40px;
`;

const CheckboxList = styled(BaseCheckboxList)`
  padding-right: 20px;
  line-height: 40px;
`;

const SelectList = styled(BaseSelectList)`
  line-height: 40px;
`;

const Results = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  padding-top: 20px;
`;

const Fact = styled(BaseFact)`
  margin-bottom: 10px;
  flex-basis: 100%;

  @media (min-width: 1024px) {
    flex-basis: calc(50% - 5px);
  }
`;

const Loading = styled(BaseLoading)`
  margin: 50px auto 0;
`;

const LAMBDA_FUNCTION_OPTIONS = [{ value: "facts", text: "Facts" }, { value: "slow-facts", text: "Slow Facts" }];
const AMOUNT_OPTIONS = [
  { value: 8, text: "8" },
  { value: 16, text: "16" },
  { value: 32, text: "32" },
  { value: 64, text: "64" }
];
const ANIMALS_OPTIONS = [
  { value: "cat", text: "Cat" },
  { value: "dog", text: "Dog" },
  { value: "snail", text: "Snail" },
  { value: "horse", text: "Horse" }
];

type Props = {
  hasError: boolean,
  isLoading: boolean,
  facts: Object[],
  amount: number,
  animals: Array<"cat" | "dog" | "snail" | "horse">,
  lambdaFunction: "facts" | "slow-facts",
  onFetchFacts: Function,
  onAmountChanged: Function,
  onLambdaFunctionChanged: Function,
  onAnimalsChanged: Function
};

class Facts extends Component<Props> {
  componentDidMount() {
    this.props.onFetchFacts();
  }

  render() {
    const {
      hasError,
      isLoading,
      facts,
      amount,
      animals,
      lambdaFunction,
      onFetchFacts,
      onAmountChanged,
      onLambdaFunctionChanged,
      onAnimalsChanged
    } = this.props;
    return (
      <Fragment>
        <header>
          <h1>Cat facts</h1>
        </header>

        <Button onClick={onFetchFacts}>Refresh Facts</Button>
        <RadioList
          title="Endpoint:"
          name="lambda"
          value={lambdaFunction}
          options={LAMBDA_FUNCTION_OPTIONS}
          onChange={onLambdaFunctionChanged}
        />
        <CheckboxList
          title="Animals:"
          name="animals"
          values={animals}
          options={ANIMALS_OPTIONS}
          onChange={onAnimalsChanged}
        />

        <SelectList title="Amount" name="amount" value={amount} options={AMOUNT_OPTIONS} onChange={onAmountChanged} />

        <Results>
          {hasError && <ErrorMessage>Error while fetching facts, try again later</ErrorMessage>}
          {isLoading && <Loading size={100} />}
          {!isLoading && facts.map(fact => <Fact key={fact._id} text={fact.text} type={fact.type} />)}
        </Results>
      </Fragment>
    );
  }
}

export default Facts;
