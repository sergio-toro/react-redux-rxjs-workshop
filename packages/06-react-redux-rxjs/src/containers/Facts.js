// @flow
import { connect } from "react-redux";

import { fetchFacts } from "../modules/facts";
import Facts from "../components/Facts";

const mapStateToProps = state => ({
  hasError: state.facts.hasError,
  isLoading: state.facts.isLoading,
  facts: state.facts.list
});

const mapDispatchToProps = {
  onFetchFacts: fetchFacts
};

const FactsContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(Facts);

export default FactsContainer;
