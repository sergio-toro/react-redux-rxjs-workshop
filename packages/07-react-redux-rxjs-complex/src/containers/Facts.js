// @flow
import { connect } from "react-redux";

import { fetchFacts, setLambdaFunction, setFactsAmount, setAnimals } from "../modules/facts";
import Facts from "../components/Facts";

const mapStateToProps = state => ({
  hasError: state.facts.hasError,
  isLoading: state.facts.isLoading,
  facts: state.facts.list,
  amount: state.facts.amount,
  animals: state.facts.animals,
  lambdaFunction: state.facts.lambdaFunction
});

const mapDispatchToProps = {
  onRefreshFacts: fetchFacts,
  onAmountChanged: setFactsAmount,
  onLambdaFunctionChanged: setLambdaFunction,
  onAnimalsChanged: setAnimals
};

const FactsContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(Facts);

export default FactsContainer;
