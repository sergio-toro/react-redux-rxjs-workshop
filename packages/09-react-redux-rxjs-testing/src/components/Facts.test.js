import React from "react";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";

import Facts from "./Facts";

describe("Facts Component", () => {
  let mockOnFetchFacts;

  beforeEach(() => {
    mockOnFetchFacts = jest.fn();
  });

  it("renders without crashing", () => {
    shallow(<Facts facts={[]} onFetchFacts={mockOnFetchFacts} />);
  });

  it("calls onFetchFacts on mount", () => {
    shallow(<Facts facts={[]} onFetchFacts={mockOnFetchFacts} />);
    expect(mockOnFetchFacts).toHaveBeenCalled();
  });

  it("it renders empty list of facts", () => {
    const output = renderer.create(<Facts facts={[]} onFetchFacts={mockOnFetchFacts} />).toJSON();
    expect(output).toMatchSnapshot();
  });

  it("it renders loading facts", () => {
    const output = renderer.create(<Facts facts={[]} isLoading={true} onFetchFacts={mockOnFetchFacts} />).toJSON();
    expect(output).toMatchSnapshot();
  });

  it("it renders a list of facts", () => {
    const list = [
      {
        _id: "the-id",
        text: "a cat is funny",
        type: "cat"
      }
    ];
    const output = renderer.create(<Facts facts={list} onFetchFacts={mockOnFetchFacts} />).toJSON();
    expect(output).toMatchSnapshot();
  });

  it("calls onFetchFacts on 'Refresh Facts' clicked", () => {
    const wrapper = shallow(<Facts facts={[]} onFetchFacts={mockOnFetchFacts} />);
    wrapper.find("RefreshButton").simulate("click");
    expect(mockOnFetchFacts).toHaveBeenCalledTimes(2);
  });
});
