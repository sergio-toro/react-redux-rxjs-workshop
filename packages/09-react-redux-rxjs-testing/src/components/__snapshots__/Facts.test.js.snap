// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Facts Component it renders a list of facts 1`] = `
Array [
  <header>
    <h1>
      Cat facts
    </h1>
  </header>,
  .c0 {
  display: inline-block;
  text-align: center;
  border-radius: 3px;
  padding: 0.5rem 0;
  margin: 0.5rem 1rem;
  width: 11rem;
  background: transparent;
  color: black;
  border: 2px solid black;
  font-family: "Segoe UI",Tahoma,Geneva,Verdana,sans-serif;
  cursor: pointer;
  margin-left: 0;
  margin-right: 20px;
}

.c0:hover {
  background: rgba(0,0,0,0.1);
}

<a
    className="c0"
    onClick={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
  >
    Refresh Facts
  </a>,
  .c1 {
  display: inline-block;
}

.c1 label {
  margin-left: 10px;
}

.c0 {
  padding-right: 20px;
  line-height: 40px;
}

<div
    className="c0 c1"
  >
    <strong>
      Endpoint:
    </strong>
    <label>
      <input
        checked={false}
        name="lambda"
        onChange={[Function]}
        type="radio"
        value="facts"
      />
        
      Facts
    </label>
    <label>
      <input
        checked={false}
        name="lambda"
        onChange={[Function]}
        type="radio"
        value="slow-facts"
      />
        
      Slow Facts
    </label>
  </div>,
  .c1 {
  display: inline-block;
}

.c1 label {
  margin-left: 10px;
}

.c0 {
  padding-right: 20px;
  line-height: 40px;
}

<div
    className="c0 c1"
  >
    <strong>
      Animals:
    </strong>
    <label>
      <input
        checked={false}
        name="animals"
        onChange={[Function]}
        type="checkbox"
        value="cat"
      />
        
      Cat
    </label>
    <label>
      <input
        checked={false}
        name="animals"
        onChange={[Function]}
        type="checkbox"
        value="dog"
      />
        
      Dog
    </label>
    <label>
      <input
        checked={false}
        name="animals"
        onChange={[Function]}
        type="checkbox"
        value="snail"
      />
        
      Snail
    </label>
    <label>
      <input
        checked={false}
        name="animals"
        onChange={[Function]}
        type="checkbox"
        value="horse"
      />
        
      Horse
    </label>
  </div>,
  .c1 {
  display: inline-block;
}

.c1 select {
  margin-left: 10px;
}

.c0 {
  line-height: 40px;
}

<label
    className="c0 c1"
  >
    <strong>
      Amount
    </strong>
    <select
      name="amount"
      onChange={[Function]}
    >
      <option
        value={8}
      >
        8
      </option>
      <option
        value={16}
      >
        16
      </option>
      <option
        value={32}
      >
        32
      </option>
      <option
        value={64}
      >
        64
      </option>
    </select>
  </label>,
  .c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-height: 75px;
  padding: 20px 30px 20px 10px;
  font-family: "Segoe UI",Tahoma,Geneva,Verdana,sans-serif;
  font-size: 20px;
  font-weight: lighter;
  line-height: 25px;
  background: #FFFAE9;
  color: #D8C685;
  border-color: #D8C685;
  border-width: 2px;
  border-style: solid;
  border-radius: 5px;
  text-shadow: 1px 1px 1px rgba(0,0,0,0.9);
  text-align: justify;
}

.c3 {
  margin-right: 20px;
  width: 100px;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding-top: 20px;
}

.c1 {
  margin-bottom: 10px;
  -webkit-flex-basis: 100%;
  -ms-flex-preferred-size: 100%;
  flex-basis: 100%;
}

@media (min-width:1024px) {
  .c1 {
    -webkit-flex-basis: calc(50% - 5px);
    -ms-flex-preferred-size: calc(50% - 5px);
    flex-basis: calc(50% - 5px);
  }
}

<div
    className="c0"
  >
    <div
      className="c1 c2"
      type="cat"
    >
      <img
        className="c3"
        src="cat.svg"
      />
      a cat is funny
    </div>
  </div>,
]
`;

exports[`Facts Component it renders empty list of facts 1`] = `
Array [
  <header>
    <h1>
      Cat facts
    </h1>
  </header>,
  .c0 {
  display: inline-block;
  text-align: center;
  border-radius: 3px;
  padding: 0.5rem 0;
  margin: 0.5rem 1rem;
  width: 11rem;
  background: transparent;
  color: black;
  border: 2px solid black;
  font-family: "Segoe UI",Tahoma,Geneva,Verdana,sans-serif;
  cursor: pointer;
  margin-left: 0;
  margin-right: 20px;
}

.c0:hover {
  background: rgba(0,0,0,0.1);
}

<a
    className="c0"
    onClick={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
  >
    Refresh Facts
  </a>,
  .c1 {
  display: inline-block;
}

.c1 label {
  margin-left: 10px;
}

.c0 {
  padding-right: 20px;
  line-height: 40px;
}

<div
    className="c0 c1"
  >
    <strong>
      Endpoint:
    </strong>
    <label>
      <input
        checked={false}
        name="lambda"
        onChange={[Function]}
        type="radio"
        value="facts"
      />
        
      Facts
    </label>
    <label>
      <input
        checked={false}
        name="lambda"
        onChange={[Function]}
        type="radio"
        value="slow-facts"
      />
        
      Slow Facts
    </label>
  </div>,
  .c1 {
  display: inline-block;
}

.c1 label {
  margin-left: 10px;
}

.c0 {
  padding-right: 20px;
  line-height: 40px;
}

<div
    className="c0 c1"
  >
    <strong>
      Animals:
    </strong>
    <label>
      <input
        checked={false}
        name="animals"
        onChange={[Function]}
        type="checkbox"
        value="cat"
      />
        
      Cat
    </label>
    <label>
      <input
        checked={false}
        name="animals"
        onChange={[Function]}
        type="checkbox"
        value="dog"
      />
        
      Dog
    </label>
    <label>
      <input
        checked={false}
        name="animals"
        onChange={[Function]}
        type="checkbox"
        value="snail"
      />
        
      Snail
    </label>
    <label>
      <input
        checked={false}
        name="animals"
        onChange={[Function]}
        type="checkbox"
        value="horse"
      />
        
      Horse
    </label>
  </div>,
  .c1 {
  display: inline-block;
}

.c1 select {
  margin-left: 10px;
}

.c0 {
  line-height: 40px;
}

<label
    className="c0 c1"
  >
    <strong>
      Amount
    </strong>
    <select
      name="amount"
      onChange={[Function]}
    >
      <option
        value={8}
      >
        8
      </option>
      <option
        value={16}
      >
        16
      </option>
      <option
        value={32}
      >
        32
      </option>
      <option
        value={64}
      >
        64
      </option>
    </select>
  </label>,
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding-top: 20px;
}

<div
    className="c0"
  />,
]
`;

exports[`Facts Component it renders loading facts 1`] = `
Array [
  <header>
    <h1>
      Cat facts
    </h1>
  </header>,
  .c0 {
  display: inline-block;
  text-align: center;
  border-radius: 3px;
  padding: 0.5rem 0;
  margin: 0.5rem 1rem;
  width: 11rem;
  background: transparent;
  color: black;
  border: 2px solid black;
  font-family: "Segoe UI",Tahoma,Geneva,Verdana,sans-serif;
  cursor: pointer;
  margin-left: 0;
  margin-right: 20px;
}

.c0:hover {
  background: rgba(0,0,0,0.1);
}

<a
    className="c0"
    onClick={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
  >
    Refresh Facts
  </a>,
  .c1 {
  display: inline-block;
}

.c1 label {
  margin-left: 10px;
}

.c0 {
  padding-right: 20px;
  line-height: 40px;
}

<div
    className="c0 c1"
  >
    <strong>
      Endpoint:
    </strong>
    <label>
      <input
        checked={false}
        name="lambda"
        onChange={[Function]}
        type="radio"
        value="facts"
      />
        
      Facts
    </label>
    <label>
      <input
        checked={false}
        name="lambda"
        onChange={[Function]}
        type="radio"
        value="slow-facts"
      />
        
      Slow Facts
    </label>
  </div>,
  .c1 {
  display: inline-block;
}

.c1 label {
  margin-left: 10px;
}

.c0 {
  padding-right: 20px;
  line-height: 40px;
}

<div
    className="c0 c1"
  >
    <strong>
      Animals:
    </strong>
    <label>
      <input
        checked={false}
        name="animals"
        onChange={[Function]}
        type="checkbox"
        value="cat"
      />
        
      Cat
    </label>
    <label>
      <input
        checked={false}
        name="animals"
        onChange={[Function]}
        type="checkbox"
        value="dog"
      />
        
      Dog
    </label>
    <label>
      <input
        checked={false}
        name="animals"
        onChange={[Function]}
        type="checkbox"
        value="snail"
      />
        
      Snail
    </label>
    <label>
      <input
        checked={false}
        name="animals"
        onChange={[Function]}
        type="checkbox"
        value="horse"
      />
        
      Horse
    </label>
  </div>,
  .c1 {
  display: inline-block;
}

.c1 select {
  margin-left: 10px;
}

.c0 {
  line-height: 40px;
}

<label
    className="c0 c1"
  >
    <strong>
      Amount
    </strong>
    <select
      name="amount"
      onChange={[Function]}
    >
      <option
        value={8}
      >
        8
      </option>
      <option
        value={16}
      >
        16
      </option>
      <option
        value={32}
      >
        32
      </option>
      <option
        value={64}
      >
        64
      </option>
    </select>
  </label>,
  .c2 {
  display: inline-block;
  position: relative;
  width: 100px;
  height: 100px;
}

.c2 div {
  position: absolute;
  border: 4px solid #000;
  opacity: 1;
  border-radius: 50%;
  -webkit-animation: lhyzzy 1s cubic-bezier(0,0.2,0.8,1) infinite;
  animation: lhyzzy 1s cubic-bezier(0,0.2,0.8,1) infinite;
}

.c2 div:nth-child(2) {
  -webkit-animation-delay: -0.5s;
  animation-delay: -0.5s;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding-top: 20px;
}

.c1 {
  margin: 50px auto 0;
}

<div
    className="c0"
  >
    <div
      className="c1 c2"
      size={100}
    >
      <div />
      <div />
    </div>
  </div>,
]
`;
